### PR TITLE
mysql* 9.0.1, mysql-connector-c++ 9.0.0

### PR DESCRIPTION
**Pull request by chenrui333**
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [x] Have you ensured that your commits follow the [commit style guide](https://docs.brew.sh/Formula-Cookbook#commit)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [ ] Have you built your formula locally with `HOMEBREW_NO_INSTALL_FROM_API=1 brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [ ] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [ ] Does your build pass `brew audit --strict <formula>` (after doing `HOMEBREW_NO_INSTALL_FROM_API=1 brew install --build-from-source <formula>`)? If this is a new formula, does it pass `brew audit --new <formula>`?

-----

### CODE REVIEW

**Comment by SMillerDev** [comment]
```
  /tmp/mysql-20240511-36861-s7mqxa/mysql-8.4.0/router/src/http/src/kdf_sha_crypt.cc:206:9: error: no matching member function for call to 'update'
        a.update({b_out.begin(), b_out.begin() + cnt});
        ~~^~~~~~
  /tmp/mysql-20240511-36861-s7mqxa/mysql-8.4.0/router/src/http/src/digest.h:94:8: note: candidate function not viable: cannot convert initializer list argument to 'const std::span<uint8_t>' (aka 'const span<unsigned char>')
    void update(const std::span<uint8_t> &data) {
         ^
  /tmp/mysql-20240511-36861-s7mqxa/mysql-8.4.0/router/src/http/src/digest.h:98:8: note: candidate function not viable: cannot convert initializer list argument to 'const std::string_view' (aka 'const basic_string_view<char>')
    void update(const std::string_view &data) {
         ^
  1 error generated.
```

**Comment by philip** [comment]
Please provide an update for this as MySQL 8.3 is a dead branch whereas MySQL 8.4 is LTS. Does the MySQL team need to fix something upstream?

**Comment by mikepage** [comment]
A new release of MySQL 8.4 addresses cmake issues:
Updated BuildRequire rules to align with versions now required for CMake and Bison. (Bug #36343254)
https://dev.mysql.com/doc/relnotes/mysql/8.4/en/news-8-4-1.html

**Comment by SMillerDev** [comment]
Let's see if 9.0.0 works now

**Comment by yahonda** [comment]
MySQL has two releases, Innovation and Long-Term Support. 
MySQL 8.4.0 is the latest LTS release and MySQL 9.0.0 is the latest innovation release. So I think some people need MySQL 8.4.0. 

Here is the blog post "Introducing MySQL Innovation and Long-Term Support (LTS) versions"
https://dev.mysql.com/blog-archive/introducing-mysql-innovation-and-long-term-support-lts-versions/

**Comment by SMillerDev** [comment]
Feel free to make a PR that adds 8.4, I'm updating this PR to track the newest release.

**Comment by philip** [comment]
I think most people that install "mysql" (e.g., "brew install mysql") want the latest LTS release, as Innovation releases can break BC. Choosing an Innovation release (e.g., 9.0.0) should be an active choice.

**Comment by SMillerDev** [comment]
> I think most people that install "mysql" (e.g., "brew install mysql") want the latest LTS release, as Innovation releases can break BC. Choosing an Innovation release (e.g., 9.0.0) should be an active choice.

And you can actively choose that by using brew and not being specific about the version. Since everything in Homebrew will install the latest version if you do not specify the version.

File: Formula/m/mysql-client@8.4.rb [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+  desc "Open source relational database management system"
+  homepage "https://dev.mysql.com/doc/refman/8.4/en/"
+  url "https://cdn.mysql.com//Downloads/MySQL-8.4/mysql-8.4.2.tar.gz"
+  sha256 "5657a78dc86bf0bf2227e0b05f8de5a2c447a816a112ffa26fa70083bcbe9814"
+  license "GPL-2.0-only" => { with: "Universal-FOSS-exception-1.0" }
+
+  livecheck do
+    url "https://dev.mysql.com/downloads/mysql/8.4.html?tpl=files&os=src&version=8.4"
+    regex(/href=.*?mysql[._-](?:boost[._-])?v?(8\.4(?:\.\d+)*)\.t/i)
+  end
```
Comment (cho-m) [inline Formula/m/mysql-client@8.4.rb:11]: Only added `mysql-client@8.4` in this PR to use as dependency. After adding `mysql@8.4` in separate PR, the livecheck can be updated.

File: Formula/p/percona-toolkit.rb [hunk 2]
```diff
@@ -27,13 +28,19 @@ class PerconaToolkit < Formula
...
+    depends_on "openssl@3"
+
+    on_intel do
+      depends_on "zstd"
+    end
+  end
+
+  on_sonoma :or_newer do
+    depends_on "zlib"
   end
```
Comment (cho-m) [inline Formula/p/percona-toolkit.rb:44]: May be worth a follow up PR to see if `-dead_strip_dylibs` can help align this odd dependency tree.

Comment (cho-m) [reply Formula/p/percona-toolkit.rb:44]: #179760

File: Formula/m/mysql-client@8.4.rb (hunk 1 above)
Comment (cho-m) [reply Formula/m/mysql-client@8.4.rb:11]: #179792

**Comment by cho-m** [comment]
May need one more run as there is going to be a conflict from rebottle.

File: Formula/m/mysql-client@8.4.rb [hunk 3]
```diff
@@ -0,0 +1,61 @@
+class MysqlClientAT84 < Formula
+  desc "Open source relational database management system"
+  homepage "https://dev.mysql.com/doc/refman/8.4/en/"
+  url "https://cdn.mysql.com//Downloads/MySQL-8.4/mysql-8.4.2.tar.gz"
```
Comment (ZhongRuoyu) [inline Formula/m/mysql-client@8.4.rb:4]: Can be done as a follow-up:

```suggestion
  url "https://cdn.mysql.com/Downloads/MySQL-8.4/mysql-8.4.2.tar.gz"
```

**Review by ZhongRuoyu — APPROVED** [review]

**Comment by Owersun** [comment]
mysql was tracking yesterday 8.3, and today it is 9.0.
If I didn't follow any beta upgrades or anything, after the "brew update" mysql is replaced to 9.0, and it doesn't start anymore with:
`2024-08-04T15:56:52.273536Z 1 [ERROR] [MY-014060] [Server] Invalid MySQL server upgrade: Cannot upgrade from 80300 to 90001. Upgrade to next major version is only allowed from the last LTS release, which version 80300 is not.`
Apparently skipping 8.4 entirely isn't the best idea.
Plus there is no mysql@8.4, only mysql-connector@8.4, and no mysql@8.3, so I can't return back to it, meaning there is no real way to start mysql anymore on top of existing database files.

**Comment by thinkverse** [comment]
For anyone who had the same issue as me and @Owersun who updated from 8.3 to 9.0.1, @simon-tma gave me a solution that worked in https://github.com/orgs/Homebrew/discussions/5539.

**Comment by nosnilmot** [comment]
`mysql@8.4` is on the way (imminently), it seems: https://github.com/Homebrew/homebrew-core/pull/179792

**Comment by carlocab** [comment]
Sorry for the pain! You should be able to
```
brew install mysql@8.4
```
now.

File: Formula/m/mysql.rb [hunk 4]
```diff
@@ -45,14 +47,14 @@ class Mysql < Formula
...
+    build 1400
+    cause "Requires C++20"
+  end
 
-  # Patch out check for Homebrew `boost`.
-  # This should not be necessary when building inside `brew`.
-  # https://github.com/Homebrew/homebrew-test-bot/pull/820
-  patch do
-    url "https://raw.githubusercontent.com/Homebrew/formula-patches/bd61f2edc4c551856f894d307140b855edb2b4f5/mysql/boost-check.patch"
-    sha256 "b90c6f78fa347cec6388d2419ee4bc9a5dc9261771eff2800d99610e1c449244"
```
Comment (carlocab) [inline Formula/m/mysql.rb:55]: Probably should've just updated this

Comment (cho-m) [reply Formula/m/mysql.rb:55]: Yeah, I realized that when working on ICU PR. Just re-added patch there. I think it usually only happens when we do mega ICU PRs.

**Pull request merged**